File: 39_combination_sum/39_combination_sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        def backTrack(remain, current, start):
            if remain < 0:
               return 
            elif remain == 0:
                result.append(list(current))
            else:
                for i in range(start, len(candidates)):
                    current.append(candidates[i])
                    backTrack(remain - candidates[i], current, i)
                    # 條件不符合所以把最後一個拿掉
                    current.pop()
        backTrack(target, [], 0)
        return result

File: 39_combination_sum/39_combination_sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function (candidates, target) {
  let result = [];

  function backTrack(remain, current, start) {
    if (remain < 0) return;
    else if (remain == 0) {
      result.push([...current]);
      return;
    } else {
      for (let i = start; i < candidates.length; i++) {
        current.push(candidates[i]);
        backTrack(remain - candidates[i], current, i);
        // 條件不符合所以把最後一個拿掉
        current.pop();
      }
    }
  }
  backTrack(target, [], 0);
  return result;
};
